geom/type_line_string.go

```diff
@@ -90,79 +91,73 @@ func (s LineString) appendWKTBody(dst []byte) []byte {
-	// tests.
-	//
-	// 1. Create slice of segments, sorted by their min X coordinate.
-	// 2. Loop over each segment.
-	//    a. Remove any elements from the heap that have their max X less than the minX of the current segment.
-	//    b. Check to see if the new element intersects with any elements in the heap.
-	//    c. Insert the current element into the heap.
-
-	if s.IsEmpty() {
+	first, last, ok := firstAndLastLines(s.seq)
+	if !ok {
 		return true
 	}
 
-	lines := make([]line, 0, s.seq.Length()-1)
-	for i := 0; i < s.seq.Length(); i++ {
+	// We need to track the index of the previous line segments, so that we can
+	// ignore the case where adjacent line segments intersect at a point (due
+	// to their construction). We can just subtract 1 from the current index,
+	// since there could be duplicated vertices in the middle of the sequence.
+	prev := -1
+
+	var tree rtree.RTree
+	n := s.seq.Length()
+	for i := 0; i < n; i++ {
 		ln, ok := getLine(s.seq, i)
-		if ok {
-			lines = append(lines, ln)
+		if !ok {
+			continue
 		}
-	}
-
-	n := len(lines)
-	unprocessed := intSequence(n)
-	sort.Slice(unprocessed, func(i, j int) bool {
-		return lines[unprocessed[i]].minX() < lines[unprocessed[j]].minX()
-	})
-
-	active := intHeap{less: func(i, j int) bool {
-		return lines[i].maxX() < lines[j].maxX()
-	}}
+		simple := true // assume simple until proven otherwise
+		box := toBox(ln.envelope())
+		tree.Search(box, func(j int) error {
+			other, ok := getLine(s.seq, j)
+			if !ok {
+				// We previously were able to access line j (otherwise we
+				// wouldn't have been able to put it into the tree). So if we
+				// can't access it now, something has gone horribly wrong.
+				panic("couldn't get line")
+			}
 
-	for _, current := range unprocessed {
-		currentX := lines[current].minX()
-		for len(active.data) != 0 && lines[active.data[0]].maxX() < currentX {
-			active.pop()
-		}
-		for _, other := range active.data {
-			inter := lines[current].intersectLine(lines[other])
+			inter := ln.intersectLine(other)
 			if inter.empty {
-				continue
+				return nil
 			}
 			if inter.ptA != inter.ptB {
 				// Two overlapping line segments.
-				return false
+				simple = false
+				return rtree.Stop
 			}
 
 			// The dimension must be 1. Since the intersection is between two
 			// lines, the intersection must be a *single* point.
 
-			if abs(current-other) == 1 {
+			if j == prev {
 				// Adjacent lines will intersect at a point due to
 				// construction, so this case is okay.
-				continue
+				return nil
 			}
 
 			// The first and last segment are allowed to intersect at a point,
 			// so long as that point is the start of the first segment and the
 			// end of the last segment (i.e. the line string is closed).
-			if (current == 0 && other == n-1) || (current == n-1 && other == 0) {
-				if s.IsClosed() {
-					continue
-				} else {
-					return false
-				}
+			atLoop := (i == first && j == last) || (i == last && j == first)
```

Comment:
Because of the way we iterate through `i`, would should be able to simplify this condition (remove one half of the OR).